File: Context/GameSpec.Base/Formats+/BinaryDds.cs
using GameSpec.Metadata;
using OpenStack.Graphics;
using OpenStack.Graphics.DirectX;
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;

// https://github.com/paroj/nv_dds/blob/master/nv_dds.cpp
namespace GameSpec.Formats
{
    public class BinaryDds : ITexture, IHaveMetaInfo
    {
        public static Task<object> Factory(BinaryReader r, FileSource f, PakFile s) => Task.FromResult((object)new BinaryDds(r));

        public BinaryDds(BinaryReader r, bool readMagic = true)
        {
            Bytes = DDS_HEADER.Read(r, readMagic, out Header, out HeaderDXT10, out Format);
            var numMipMaps = Math.Max(1, Header.dwMipMapCount);
            var offset = 0;
            Mips = new Range[numMipMaps];
            for (var i = 0; i < numMipMaps; i++)
            {
                int w = (int)Header.dwWidth >> i, h = (int)Header.dwHeight >> i;
                if (w == 0 || h == 0) { Mips[i] = -1..; continue; }
                var size = ((w + 3) / 4) * ((h + 3) / 4) * Format.blockSize;
                var remains = Math.Min(size, Bytes.Length - offset);
                Mips[i] = remains > 0 ? offset..(offset + remains) : -1..;
                offset += remains;
            }
        }

        DDS_HEADER Header;
        DDS_HEADER_DXT10? HeaderDXT10;
        (object type, int blockSize, object gl, object vulken, object unity, object unreal) Format;
        byte[] Bytes;
        Range[] Mips;

        public IDictionary<string, object> Data => null;
        public int Width => (int)Header.dwWidth;
        public int Height => (int)Header.dwHeight;
        public int Depth => 0;
        public int MipMaps => (int)Header.dwMipMapCount;
        public TextureFlags Flags => 0;

        public byte[] Begin(int platform, out object format, out Range[] mips)
        {
            format = (FamilyPlatform.Type)platform switch
            {
                FamilyPlatform.Type.OpenGL => Format.gl,
                FamilyPlatform.Type.Unity => Format.unity,
                FamilyPlatform.Type.Unreal => Format.unreal,
                FamilyPlatform.Type.Vulken => Format.vulken,
                FamilyPlatform.Type.StereoKit => throw new NotImplementedException("StereoKit"),
                _ => throw new ArgumentOutOfRangeException(nameof(platform), $"{platform}"),
            };
            mips = Mips;
            return Bytes;
        }
        public void End() { }

        List<MetaInfo> IHaveMetaInfo.GetInfoNodes(MetaManager resource, FileSource file, object tag) => new List<MetaInfo> {
            new MetaInfo(null, new MetaContent { Type = "Texture", Name = Path.GetFileName(file.Path), Value = this }),
            new MetaInfo("Texture", items: new List<MetaInfo> {
                new MetaInfo($"Format: {Format.type}"),
                new MetaInfo($"Width: {Width}"),
                new MetaInfo($"Height: {Height}"),
                new MetaInfo($"Mipmaps: {MipMaps}"),
            }),
        };
    }
}

File: Python/gamespecs/Base/binary_dds.py
import os
from io import BytesIO
from typing import Any
from openstk.poly import Reader
from openstk.gfx_dds import DDS_HEADER
from openstk.gfx_texturemgr import TextureFlags
from ..pakfile import FileSource, PakFile
from ..metamgr import MetaManager, MetaInfo, MetaContent, IHaveMetaInfo

class Binary_Dds(IHaveMetaInfo):
    @staticmethod
    def factory(r: Reader, f: FileSource, s: PakFile): return Binary_Dds(r)

    def __init__(self, r: Reader, readMagic: bool = True):
        self.bytes, self.header, self.headerDXT10, self.format = DDS_HEADER.read(r, readMagic)
        numMipMaps = max(1, self.header.dwMipMapCount)
        offset = 0
        self.mips = [range(-1, 0)] * numMipMaps
        for i in range(numMipMaps):
            w = self.header.dwWidth >> i; h = self.header.dwHeight >> i
            if w == 0 or h == 0: self.mips[i] = range(-1, 0); continue
            size = int(((w + 3) / 4) * ((h + 3) / 4)) * self.format[1]
            remains = min(size, len(self.bytes) - offset)
            self.mips[i] = range(offset, (offset + remains)) if remains > 0 else range(-1, 0)
            offset += remains
    def data(self) -> dict[str, object]: return None
    def width(self) -> int: return self.header.dwWidth
    def height(self) -> int: return self.header.dwHeight
    def depth(self) -> int: return 0
    def mipMaps(self) -> int: return self.header.dwMipMapCount
    def flags(self) -> TextureFlags: return 0

    def begin(self, platform: int) -> (bytes, object, list[Any]):
        match platform:
            case FamilyPlatform.Type.OpenGL: format = Format.gl
            case FamilyPlatform.Type.Unity: format = Format.unity
            case FamilyPlatform.Type.Unreal: format = Format.unreal
            case FamilyPlatform.Type.Vulken: format = Format.vulken
            case _: raise Exception('Unknown {platform}')
        return self.bytes, format, self.mips
    def end(self): pass

    def getInfoNodes(self, resource: MetaManager = None, file: FileSource = None, tag: object = None) -> list[MetaInfo]: return [
        MetaInfo(None, MetaContent(type = 'Texture', name = os.path.basename(file.path), value = self)),
        MetaInfo('Texture', items = [
            MetaInfo(f'Format: {self.format[0]}'),
            MetaInfo(f'Width: {self.width()}'),
            MetaInfo(f'Height: {self.height()}'),
            MetaInfo(f'Mipmaps: {self.mipMaps()}')
            ])
        ]
